### PR TITLE
Run on 2016+2017+2018

### PR DESCRIPTION
**Pull request by francescobrivio**
Adding all necessary for running on three years.
Useful informations from:
https://twiki.cern.ch/twiki/bin/view/CMS/HiggsToTauTauWorkingLegacyRun2

### CODE REVIEW

**Comment by francescobrivio** [comment]
@camendola please wait to merge this, I'm still adding more commits

**Comment by francescobrivio** [comment]
Preliminary Data and Backgrounds for 2016+2017+2018 added.
Still need to check:
- which "others" backgrounds we want to add
- fix the "singleHiggs" backgrounds
- choose only one from the "extended" version
- add the non-resonant signals:
  - Gluon fusion
  - VBF

File: NtupleProducer/test/datasets_Legacy2016.txt [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+/SingleMuon/Run2016B-17Jul2018_ver2-v1/MINIAOD
+/SingleMuon/Run2016C-17Jul2018-v1/MINIAOD
+/SingleMuon/Run2016D-17Jul2018-v1/MINIAOD
+/SingleMuon/Run2016E-17Jul2018-v1/MINIAOD
+/SingleMuon/Run2016F-17Jul2018-v1/MINIAOD
+/SingleMuon/Run2016G-17Jul2018-v1/MINIAOD
+/SingleMuon/Run2016H-17Jul2018-v1/MINIAOD
+
+/Tau/Run2016B-17Jul2018_ver1-v1/MINIAOD
+/Tau/Run2016B-17Jul2018_ver1-v2/MINIAOD
```
Comment (camendola) [inline NtupleProducer/test/datasets_Legacy2016.txt]: -> /Tau/Run2016B-17Jul2018_ver2-v1/MINIAOD

File: NtupleProducer/test/datasets_Legacy2016.txt [hunk 2]
```diff
@@ -0,0 +1,96 @@
+=== DATA_2016 ===
+/SingleElectron/Run2016B-17Jul2018_ver1-v1/MINIAOD
+/SingleElectron/Run2016B-17Jul2018_ver2-v1/MINIAOD
+/SingleElectron/Run2016C-17Jul2018-v1/MINIAOD
+/SingleElectron/Run2016D-17Jul2018-v1/MINIAOD
+/SingleElectron/Run2016E-17Jul2018-v1/MINIAOD
+/SingleElectron/Run2016F-17Jul2018-v1/MINIAOD
+/SingleElectron/Run2016G-17Jul2018-v1/MINIAOD
+/SingleElectron/Run2016H-17Jul2018_ver2-v1/MINIAOD
+/SingleElectron/Run2016H-17Jul2018_ver3-v1/MINIAOD
```
Comment (camendola) [inline NtupleProducer/test/datasets_Legacy2016.txt]: there's only /SingleElectron/Run2016H-17Jul2018-v1/MINIAOD

File: NtupleProducer/test/datasets_Legacy2016.txt [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+/Tau/Run2016B-17Jul2018_ver1-v1/MINIAOD
+/Tau/Run2016B-17Jul2018_ver1-v2/MINIAOD
+/Tau/Run2016C-17Jul2018-v1/MINIAOD
+/Tau/Run2016D-17Jul2018-v1/MINIAOD
+/Tau/Run2016E-17Jul2018-v1/MINIAOD
+/Tau/Run2016F-17Jul2018-v1/MINIAOD
+/Tau/Run2016G-17Jul2018-v1/MINIAOD
+/Tau/Run2016H-17Jul2018_ver2-v1/MINIAOD
+/Tau/Run2016H-17Jul2018_ver3-v1/MINIAOD
```
Comment (camendola) [inline NtupleProducer/test/datasets_Legacy2016.txt]: there is only /Tau/Run2016H-17Jul2018-v1/MINIAOD

File: NtupleProducer/test/datasets_Legacy2016.txt (hunk 1 above)
Comment (francescobrivio) [reply NtupleProducer/test/datasets_Legacy2016.txt]: should this replace both /Tau/Run2016B or just the second one?

Comment (camendola) [reply NtupleProducer/test/datasets_Legacy2016.txt]: only the second one

**Review by dzuolo — COMMENTED** [review]
Hi! On the trigger list of 2018 did you check this?

Dear all,

We would like to bring your attention to different muon filters used in 2018 data taking. Thanks a lot Janek for pointing this out and sorry for the late announcement to you. 

In the first 5.3/fb data was taken with a muon filter “hltL3crIsoL1sMu18erTau24erIorMu20erTau24erL1f0L2f10QL3f20QL3trkIsoFiltered0p07”, while it was changed in HLT Menu v2.0 started in Run 315974.

In the remaining part of 2018 for Run > = 315974, the muon filter used was “hltL3crIsoBigORMu18erTauXXer2p1L1f0L2f10QL3f20QL3trkIsoFiltered0p07”.

We updated the tau trigger twiki correspondingly: https://twiki.cern.ch/twiki/bin/viewauth/CMS/TauTrigger#Trigger_table_for_2018

Please also note the changes in the name of the trigger paths and filters for HPS as noted in the twiki.

**Comment by francescobrivio** [comment]
@dzuolo thanks I missed that!
I now fixed by duplicating (with updated filters) the muTau triggers.
Since there were too many HLT paths (stored in 64 bits) I commented the "medium" and "tight" hlt paths (in KLUB we use only the "loose" one that is inclusive of the others)

**Comment by francescobrivio** [comment]
I just realized that the way I implemented the new filters for MuTau triggers is not correct.
@l-cadamuro do you have an idea how to solve this? You can read more on this here: https://github.com/LLRCMS/LLRHiggsTauTau/pull/103#pullrequestreview-330257968

**Comment by francescobrivio** [comment]
Finally fixed the MuTau filters!
Changes made:
- 102X_triggers are back to original file --> these are the default triggers (to be used also for MC)
- if running on 2018 data with run<315974 the triggerMap of the MuTau filters are custmized with the correct filters (this happens in the Ntuplizer)
- One function added in triggerMapper.h/cc and two functions added in triggerHelper.h/cc in order to properly customize the map with the new filters

Tested on Data and MC and everything seems to work fine.
@camendola , @dzuolo and  @l-cadamuro if you have time please have a look at the last commit ( https://github.com/LLRCMS/LLRHiggsTauTau/pull/103/commits/1126da4b324fec509b4051669db5007df248b2a5 )
to check if it's ok with you. Thanks!

**Comment by francescobrivio** [comment]
Added L1ECALPrefiring following the twiki:
https://twiki.cern.ch/twiki/bin/viewauth/CMS/L1ECALPrefiringWeightRecipe
Three branches added with the prefire weight per event.
The weights are computed for 2016 and 2017 MC only, in the other cases the branches are filled with 1.0.

**Pull request merged**